scripts/irods/upgrade_configuration.py

```diff
@@ -33,6 +33,7 @@ def requires_upgrade(irods_config):
     new_version_tuple = lib.version_string_to_tuple(new_version['irods_version'])
     old_version_tuple = lib.version_string_to_tuple(irods_config.version['irods_version'])
 
+    # TODO Thoughts on changing this from "!=" to "<". Using not-equals feels incorrect.
```

Comment:
Yes. That's what I was thinking about when writing that comment.

Can you elaborate? It's not clear to me what your thoughts are on this.

Reply:
someone could have upgraded their system and now running an 'older' server.... aka we would detect this as a 'downgrade'.

so if we limit this to be '<', then we'd miss that scenario?

Reply:
I see. How important is it to detect and report downgrade cases?

In iRODS 4, the use of `!=` allows the `upgrade` path to be triggered, eventually hitting the downgrade detection logic. This works.

In this PR, the use of `<` does not allow the `upgrade` path to be triggered at all, removing the need for downgrade detection logic too.

Keep in mind that the upgrade logic for iRODS 4 was executed on server startup where as in iRODS 5, the admin must run `./irodsctl upgrade` before starting the server.

Reply:
right - will need to think through the scenarios a bit more.

we definitely cannot support downgrades, because we reserve the right to change the data in the database, and a downgrade would not be able to 'know' what the values were 'before'.

as long as we can detect downgrades and give good error messages, then I think everything else is fine.

Reply:
and yes, this is localized to this 'requires_upgrade()' function... so using '<' is probably fine here.   but the remaining logic should still detect downgrade and help the admin understand what is happening.

Reply:
Yeah, I think separating the two paths is the best approach.

I'll modify `./irodsctl upgrade` to do the following:
1. Check for downgrade first. If downgrading, print a helpful error message.
2. Check for upgrades and apply accordingly.

Reply:
i like it.

clean and exits early before getting into any trouble.

Reply:
Initial implementation pushed.

It has not been tested. While I do feel it's better, I don't know if it works for 4.0 or 4.1 server. I replicated logic from other parts of the codebase so I believe it's correct.

If we're uncomfortable with this, I see to paths forward for keeping it:
- (A) Declare that iRODS 5 only supports upgrading from 4.2
- (B) We go back to doing what the original code did (i.e. revert back to using `!=` and letting the upgrade function check for downgrades)

Reply:
this will only be in 5... so... i think we'll be fine.  we'll get it all tested.